source/Nevermore/SelectBuilder.cs

```diff
@@ -0,0 +1,232 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nevermore.AST;
+
+namespace Nevermore
+{
+    public class JoinSelectBuilder : SelectBuilderBase<JoinedSource>
+    {
+        public JoinSelectBuilder(JoinedSource @from) : this(@from, new List<IWhereClause>(), new List<OrderByField>())
+        {
+        }
+
+        JoinSelectBuilder(JoinedSource @from, List<IWhereClause> whereClauses, List<OrderByField> orderByClauses) 
+            : base(@from, whereClauses, orderByClauses)
+        {
+        }
+
+        protected override ISelectColumns DefaultSelect => new SelectAllFrom(From.Source.Alias);
+
+        protected override IEnumerable<OrderByField> GetDefaultOrderByFields()
+        {
+            yield return new OrderByField(new TableColumn(new Column("Id"), From.Source.Alias));
+        }
+
+        public override void AddWhere(UnaryWhereParameter whereParams)
+        {
+            WhereClauses.Add(new UnaryWhereClause(new AliasedWhereFieldReference(From.Source.Alias, new WhereFieldReference(whereParams.FieldName)), 
```

Comment:
You've built a nice immutable AST, but then suddenly here we fall back into a mutable list. I think we should go one step further and make a Query AST object that encapsulates the whole query (Select, From, Where and OrderBy). This single root object is then wrapped by the QueryBuilder to provide a nicer interface/backwards compatibility.

Reply:
That's basically what the `ISelectBuilder` is. It's not part of the AST. It has mutable objects that represent Where, OrderBy and Select. From is a slightly more complicated so happens to be immutable. 

Maybe I'm not understanding you correctly.

Reply:
Perhaps I missed it in the code, but is there an object in the AST that represents the Query itself? i.e I can get a reference to that object and see the whole query end to end?

Reply:
Yep, the root of the AST is the `Select` class.